feat: canvas viewer

bbqben opened the pull request:
# PR Description

## Summary of Changes

- **Introduced a new Canvas feature** in the renderer, providing a sidebar within the chat interface to view files.
- Added components for managing open documents in tabs (`CanvasTabs`), displaying content (`CanvasContent`), rendering code (`CodeEditor`), and previewing markdown (`MarkdownPreview`).
- Implemented a **Zustand store (`canvas-store.ts`)** to manage the state of the canvas, including open documents and the active document.
- Added a new **`readFile` IPC handler** in the main process and exposed it via the preload script to allow the renderer to read file content.
- Integrated the `Canvas` component into the `ChatContent` using a resizable split panel.
- Updated the `ChatHeader` to include an "Open Canvas" button.
- Modified the `FileAttachment` component to open supported files directly in the canvas.
- Added utility functions for checking canvas support, identifying markdown files, and loading CodeMirror language extensions.
- Added new dependencies for CodeMirror and react-split.

## Related Issue(s)

- Related: #...

## Impact

- **Enhances user experience** by allowing direct file viewing within the application.
- **Adds support for viewing markdown and code files** with syntax highlighting.
- Introduces a **resizable sidebar panel** for the canvas.
- **No known breaking changes** to existing functionality.
- Adds new dependencies for the canvas feature.
- **No known performance or security implications** beyond the addition of new features and dependencies.

## Testing Details

- Verified that the canvas opens and closes correctly.
- Tested opening various file types (markdown, code files) from chat messages.
- Confirmed that markdown files are rendered correctly in the preview.
- Confirmed that code files are displayed with syntax highlighting in the editor.
- Tested opening multiple files and switching between tabs.
- Verified that closing tabs removes the corresponding document from the canvas.
- Ensured the split panel is resizable.
- Checked console for any errors related to file reading or component rendering.

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my code.
- [ ] I have added tests that prove my changes work as intended. (Note: Automated tests for UI components and IPC handlers may be needed)
- [ ] All new and existing tests pass. (Note: Existing tests should pass, new tests need to be added)
- [x] I have run formatting (biome), linting, and type checks, and they pass.
- [ ] I have updated the documentation when required. (Note: Documentation for the new feature may be needed)
- [x] Security considerations are addressed (especially for code execution features). (Note: File reading is handled via a secure IPC channel)
- [x] I have linked all related issues.

## Screenshots (Optional)

If applicable, please attach screenshots that illustrate the changes or new features.


https://github.com/user-attachments/assets/a23bb0a4-e31b-4fb1-b468-0955494e8f50


https://github.com/user-attachments/assets/aef8c463-7ecc-460e-b295-a28e2f693fd5



## Summary

- This PR introduces a **new Canvas feature** for viewing files directly within the chat interface.
- It includes new UI components, state management, IPC communication for file reading, and utility functions.
- The feature enhances the user experience by providing integrated file viewing capabilities.

damianvtran commented:
This is amazing, so much easier to work with docs now 🙏 🔥 

Some small stuff:
1. Overflow on longer file names on tabs, maybe we can truncate to some fixed length of chars (32?) with ellipsis at the beginning of the string, need to compare to how this is done in vscode, etc.


<img width="255" alt="image" src="https://github.com/user-attachments/assets/d6ea2b6e-2f78-4b96-8394-d1dadab9f87f" />


2. There's a white border on darker themes around the main chat area and the canvas that maybe needs to become theme-aware.  Probably, we should reduce the border width?


<img width="938" alt="image" src="https://github.com/user-attachments/assets/17c9e1cd-55b5-48a0-9777-1b678efee9ba" />


3. Looks like the scroll to bottom button got centered, I think the button in general needs to be improved but for now, probably just check to see how new parent styles might have affected it and restore it to bottom-right


<img width="959" alt="image" src="https://github.com/user-attachments/assets/f1d72f11-6415-4c44-85e1-9fb5e5325f19" />


4. Can probably apply the same webkit scrollbar themes as some of the other components (see the messages view as a reference probably) to make the scrollbar blend in nicely


Canvas:


<img width="77" alt="image" src="https://github.com/user-attachments/assets/fa4cd14f-0e8f-4976-9907-9d6c8fe97679" />


Messages view:

<img width="105" alt="image" src="https://github.com/user-attachments/assets/ba792930-e63d-4a15-9056-a08ea71c6cfc" />


5. [Feature Request] Would be cool to have copy and export options - copy to copy the doc to paste it into other tools and export with pandoc or others to save as a PDF/docx and share the docs with others.  Not a blocker at all to merge this one in but food for thought!

bbqben commented:
> 1. Overflow on longer file names on tabs, maybe we can truncate to some fixed length of chars (32?) with ellipsis at the beginning of the string, need to compare to how this is done in vscode, etc.

In VSCode it actually renders the full title, so in this case what we can do is remove the max-width that's currently set it just allow for it to stretch as long as it needs to. 

<img width="697" alt="image" src="https://github.com/user-attachments/assets/8472ffee-b4d4-4f08-b8a6-5dfd513b529c" />

<br/>


> 2. There's a white border on darker themes around the main chat area and the canvas that maybe needs to become theme-aware.  Probably, we should reduce the border width?

mb this was actually left in from either debugging or copied from the example. I've removed it in this commit, https://github.com/damianvtran/local-operator-ui/pull/38/commits/9633f32f07eb52465bb75241becf18c25fe8d6c7

<br/>

> 3. Looks like the scroll to bottom button got centered, I think the button in general needs to be improved but for now, probably just check to see how new parent styles might have affected it and restore it to bottom-right

So this was actually intentional and was done in this commit, https://github.com/damianvtran/local-operator-ui/pull/38/commits/8b828cc0979f7336af9f40ab493329320fe877c5. The reason behind it is that on smaller widths it goes on top of the message input which doesn't feel right having it block interactive elements. 
<img width="819" alt="image" src="https://github.com/user-attachments/assets/5f8494a9-9f1e-4e46-8754-3bd602847968" />

The new location being above the message box matches similarly to how OpenAI has it, 
<img width="1044" alt="image" src="https://github.com/user-attachments/assets/a7196680-6e63-4a2e-bae8-e1b7b22454c3" />

Comparing it side-by-side it stands out a lot more compared to OpenAI and so we can make the sizing of it smaller and remove the box shadow, but keeping it close to the chat window keeps it in view so the user can associate it to the chat. I think the association is not as strong having it in the corner. Alternatively we can have two different styles depending on the width of the screen

<br/>

> 4. Can probably apply the same webkit scrollbar themes as some of the other components (see the messages view as a reference probably) to make the scrollbar blend in nicely

I've updated the styles to have it match the ones from messages view. Also I think we'd probably want to set these styles globally on the application, there's quite a few of these styles just repeating around the codebase. I'll add this as a list of items to tackle

<br/>

> 5. [Feature Request] Would be cool to have copy and export options - copy to copy the doc to paste it into other tools and export with pandoc or others to save as a PDF/docx and share the docs with others.  Not a blocker at all to merge this one in but food for thought!

Ya this one I'm planning to tackle as a follow up. There was some work done to setup a drawer and add download/export functionality but I left it out for the time being ([reference here](https://github.com/damianvtran/local-operator-ui/pull/38/files#diff-f2ad9fc1b7d960894dda3d03587e2e71c44a26cac09532e2709904323818dab4)). Wanted to confirm some requirements details before continuing to work on this portion.

damianvtran commented:
Made some small touch-ups for QOL and styling

Use theme border when not hovering
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/6b9f62df-a8b4-4a12-9319-e05bfb994ba5" />

When you bring the mouse into the hover zone, the border widens and a handle appears
<img width="287" alt="image" src="https://github.com/user-attachments/assets/63c34e88-38ff-4fc3-a4b9-a7d18e996532" />

Implemented canvas and ui preferences store to keep track of the canvas open status and which files are open in the canvas already so that the user will be able to get back to them if they switch conversations/refresh

damianvtran commented:
The UI preferences store is also used to preserve the canvas width

I also created a separate divider component and applied the same logic to the left chat sidebar so both sides are now configurable in width
